File: nim/GameOfLife.nim
import random
import std/strformat

const
  generations = 500
  height = 50
  width = 100

type Board = object
  cells: seq[bool]
  height: int
  width: int

proc getCell(board: Board, row: int, col: int): bool =
  if row >= 0 and col >= 0 and row < board.height and col < board.width:
    board.cells[(row * board.width) + col]
  else:
    false

proc nextGeneration(board: var Board) =
  var boardCopy = Board(height: height, width: width)
  for i in countup(0, (board.height*board.width)-1):
    boardCopy.cells.add(board.cells[i])

  for i in countup(0, height-1):
    for j in countup(0, width-1):
      var cell = getCell(boardCopy, i, j)
      var adjacent = 0

      for n in countup(-1, 1):
        for m in countup(-1, 1):
          if (n == 0 and m == 0):
            continue
          if getCell(boardCopy, i+n, j+m):
            adjacent += 1
      
      if cell:
        if adjacent < 2:
          cell = false
        if adjacent > 3:
          cell = false
      else:
        if adjacent == 3:
          cell = true

      board.cells[(i*board.width)+j] = cell

proc drawScreen(board: Board) =
  for i in countup(0, board.height-1):
    for j in countup(0, board.width-1):
      if board.cells[(i*board.width)+j]:
        write(stdout, "██")
      else:
        write(stdout, "  ")
    writeLine(stdout, "")

proc runGame(height: int, width: int, generations: int) =
  var board = Board(height: height, width: width)

  randomize()
  for i in countup(0, (height * width)-1):
    board.cells.add(sample([true, false]))

  for i in countup(0, generations-1):
    drawScreen(board)
    write(stdout, &"\x1b[{board.height}A")
    nextGeneration(board)

runGame(height, width, generations)

File: lua/game_of_life.lua
generations = 500
height = 50
width = 100

function get_cell(board, row, col)
    if row >= 1 and col >= 1 and row <= board.height and col <= board.width then
        return board.cells[(row*board.width)+col]
    else
        return false
    end
end

function next_generation(board)
    board_copy = {}
    board_copy.height = board.height
    board_copy.width = board.width
    board_copy.cells = {}
    for k,v in pairs(board.cells) do
        board_copy.cells[k] = v
    end

    for i=1,board.height do
        for j=1,board.width do
            cell = get_cell(board_copy, i, j)
            adjacent = 0

            for n=-1,1 do
                for m=-1,1 do
                    if not ((n == -1 and i == 1) or (m == -1 and j == 1) or (n == 0 and m == 0)) then
                        if get_cell(board_copy, i+n, j+m) then
                            adjacent = adjacent + 1
                        end
                    end
                end
            end
                        
            if cell then
                if adjacent < 2 then
                    cell = false
                end
                if adjacent > 3 then
                    cell = false
                end
            else
                if adjacent == 3 then
                    cell = true
                end
            end

            board.cells[(i*board.width)+j] = cell
        end
    end
end

function draw_screen(board)
    for i=1,board.height do
        for j=1,board.width do
            if board.cells[(i*board.width)+j] then
                io.write("██")
            else
                io.write("  ")
            end
        end

        io.write("\n")
    end
end

function run_game(height, width, generations)
    board = {}
    board.height = height
    board.width = width
    board.cells = {}
    for i=1,height*width do
        local trueAndFalse = {true, false}
        board.cells[i] = trueAndFalse[math.random(1, 2)]
    end

    for i=1,generations do
        draw_screen(board)
        io.write(string.char(0x1b) .. "[" .. board.height .. "A")
        next_generation(board)
    end
end

run_game(height, width, generations)